blas/gonum/level3cmplx128.go

```diff
@@ -254,6 +254,173 @@ func (Implementation) Zgemm(tA, tB blas.Transpose, m, n, k int, alpha complex128
 	}
 }
 
+// Zherk performs one of the hermitian rank-k operations
+//  C = alpha*A*A^H + beta*C  if trans == blas.NoTrans
+//  C = alpha*A^H*A + beta*C  if trans == blas.ConjTrans
+// where alpha and beta are real scalars, C is an n×n hermitian matrix and A is
+// an n×k matrix in the first case and a k×n matrix in the second case.
+//
+// The imaginary parts of the diagonal elements of C are assumed to be zero, and
+// on return they will be set to zero.
+func (Implementation) Zherk(uplo blas.Uplo, trans blas.Transpose, n, k int, alpha float64, a []complex128, lda int, beta float64, c []complex128, ldc int) {
+	var rowA, colA int
+	switch trans {
+	default:
+		panic(badTranspose)
+	case blas.NoTrans:
+		rowA, colA = n, k
+	case blas.ConjTrans:
+		rowA, colA = k, n
+	}
+	switch {
+	case uplo != blas.Lower && uplo != blas.Upper:
+		panic(badUplo)
+	case n < 0:
+		panic(nLT0)
+	case k < 0:
+		panic(kLT0)
+	case lda < max(1, colA):
+		panic(badLdA)
+	case ldc < max(1, n):
+		panic(badLdC)
+	}
+
+	// Quick return if possible.
+	if n == 0 {
+		return
+	}
+
+	// For zero matrix size the following slice length checks are trivially satisfied.
+	if len(a) < (rowA-1)*lda+colA {
+		panic(shortA)
+	}
+	if len(c) < (n-1)*ldc+n {
+		panic(shortC)
+	}
+
+	// Quick return if possible.
+	if (alpha == 0 || k == 0) && beta == 1 {
+		return
+	}
+
+	if alpha == 0 {
+		if uplo == blas.Upper {
+			if beta == 0 {
+				for i := 0; i < n; i++ {
+					ci := c[i*ldc+i : i*ldc+n]
+					for j := range ci {
+						ci[j] = 0
+					}
+				}
+			} else {
+				for i := 0; i < n; i++ {
+					ci := c[i*ldc+i : i*ldc+n]
+					ci[0] = complex(beta*real(ci[0]), 0)
+					if i != n-1 {
+						c128.DscalUnitary(beta, ci[1:])
+					}
+				}
+			}
+		} else {
+			if beta == 0 {
+				for i := 0; i < n; i++ {
+					ci := c[i*ldc : i*ldc+i+1]
+					for j := range ci {
+						ci[j] = 0
+					}
+				}
+			} else {
+				for i := 0; i < n; i++ {
+					ci := c[i*ldc : i*ldc+i+1]
+					if i != 0 {
+						c128.DscalUnitary(beta, ci[:i])
+					}
+					ci[i] = complex(beta*real(ci[i]), 0)
+				}
+			}
+		}
+		return
+	}
+
+	calpha := complex(alpha, 0)
+	if trans == blas.NoTrans {
+		// Form  C = alpha*A*A^H + beta*C.
+		cbeta := complex(beta, 0)
+		if uplo == blas.Upper {
+			for i := 0; i < n; i++ {
```

Comment:
Cases for beta==0?

Reply:
I'll add them.

Reply:
This is important because of NaN behaviour.

Reply:
I split it anyway, I think it's clearer this way. Also added a more descriptive message to the fixing commit.